datafusion/core/src/physical_plan/joins/nested_loop_join.rs

```diff
@@ -479,6 +404,185 @@ impl NestedLoopJoinStream {
+            "Build left right indices error".to_string(),
+        )),
+        Ok(indices) => {
+            for (left_side, right_side) in indices {
+                left_indices_builder
+                    .append_values(left_side.values(), &vec![true; left_side.len()]);
+                right_indices_builder
+                    .append_values(right_side.values(), &vec![true; right_side.len()]);
+            }
+            Ok((
+                left_indices_builder.finish(),
+                right_indices_builder.finish(),
+            ))
+        }
+    };
+    match left_right_indices {
+        Ok((left_side, right_side)) => {
+            // set the left bitmap
+            // and only full join need the left bitmap
+            if join_type == JoinType::Full {
+                left_side.iter().flatten().for_each(|x| {
+                    visited_left_side.set_bit(x as usize, true);
+                });
+            }
+            // adjust the two side indices base on the join type
+            let (left_side, right_side) = adjust_indices_by_join_type(
+                left_side,
+                right_side,
+                left_batch.num_rows(),
+                right_batch.num_rows(),
+                join_type,
+            );
+
+            build_batch_from_indices(
+                schema,
+                left_batch,
+                right_batch,
+                left_side,
+                right_side,
+                column_indices,
+            )
+        }
+        Err(e) => Err(e),
+    }
+}
+
+fn adjust_indices_by_join_type(
+    left_indices: UInt64Array,
+    right_indices: UInt32Array,
+    count_left_batch: usize,
+    count_right_batch: usize,
+    join_type: JoinType,
+) -> (UInt64Array, UInt32Array) {
+    match join_type {
+        JoinType::Inner => (left_indices, right_indices),
+        JoinType::Left => {
+            // matched
+            // unmatched left row will be produced in this batch
+            let left_unmatched_indices =
+                get_anti_u64_indices(count_left_batch, &left_indices);
+            // combine the matched and unmatched left result together
+            append_left_indices(left_indices, right_indices, left_unmatched_indices)
+        }
+        JoinType::LeftSemi => {
+            // need to remove the duplicated record in the left side
+            let left_indices = get_semi_u64_indices(count_left_batch, &left_indices);
+            // the right_indices will not be used later for the `left semi` join
+            (left_indices, right_indices)
+        }
+        JoinType::LeftAnti => {
+            // need to remove the duplicated record in the left side
+            // get the anti index for the left side
+            let left_indices = get_anti_u64_indices(count_left_batch, &left_indices);
+            // the right_indices will not be used later for the `left anti` join
+            (left_indices, right_indices)
+        }
```

Comment:
Left/LeftSemi/LeftAnti joins do not need on bitmap because the right is single partition.